typeset/converter.js

```diff
@@ -129,73 +146,44 @@ const createMapOfMathMLElements = async (log, inputPath, cssPath, outputPath, ou
       index++
     }
     return [...mathMap.entries()]
-  }, PROGRESS_TIME)
-
-  const convertedMathML/*: Map<string, {svg, html, css}> */ = await mjnodeConverter.convertMathML(log, new Map(mathEntries), outputFormat)
+  }, PROGRESS_TIME, mathNodePlaceholder('{id}'))
 
-  log.info(`Inserting converted math elements...`)
-  const mathSources = [...convertedMathML.entries()]
-    .map(([id, {svg, html}]) => [id, svg || html])
-  await page.evaluate((convertedMathMLEntries, PROGRESS_TIME) => {
-    const total = convertedMathMLEntries.length
-    let prevTime = Date.now()
-    let index = 0
-    for (const [id, xml] of convertedMathMLEntries) {
-      const mathHTML = xml
-      const mathNode = document.getElementById(id)
-      if (!mathNode) {
-        throw new Error(`BUG: Could not find element with id="${id}"`)
-      }
-      try {
-        mathNode.outerHTML = mathHTML
-      } catch (err) {
-        console.error(`Problem inserting id="${id}" back into the document (might not be valid XML)`)
-        console.error(mathHTML)
-        mathNode.outerHTML = mathHTML
-      }
+  let pageContent = await page.evaluate(() => {
+    const serializer = new window.XMLSerializer()
+    return serializer.serializeToString(document.documentElement)
+  })
 
-      // Print progress every 10 seconds
-      const now = Date.now()
-      if (now - prevTime > PROGRESS_TIME) {
-        const percent = Math.round(100 * index / total)
-        console.info(`Inserted ${percent}% of all elements...`)
-        prevTime = now
-      }
-      index++
+  browser.close()
+
+  const BATCH_SIZE = 3000
+  let allUniqueCss = new Set();
+  let nextMathNodeIndex
+  for (let batch = 0; batch < Math.ceil(mathEntries.length / BATCH_SIZE); batch++) {
+    const start = BATCH_SIZE * batch
+    const end = Math.min(BATCH_SIZE * batch + BATCH_SIZE, mathEntries.length)
+    log.info(`Converting math elements ${start} to ${end} of ${mathEntries.length}`)
+    const [convertedMathML/*: Map<string, {svg || html}> */, uniqueCss] = await mjnodeConverter.convertMathML(log, new Map(mathEntries.slice(start, end)), outputFormat, mathEntries.length, start)
+    allUniqueCss.add(uniqueCss)
+
+    log.debug(`Inserting converted math elements...`)
+    let convertedContent = []
+    for (const [id, xml] of convertedMathML) {
+      // Replacing placeholder with converted math
+      nextMathNodeIndex = pageContent.indexOf(mathNodePlaceholder(id))
+      convertedContent.push(pageContent.substr(0, nextMathNodeIndex))
+      convertedContent.push(xml)
+      pageContent = pageContent.substr(nextMathNodeIndex + mathNodePlaceholder(id).length)
     }
-  }, mathSources, PROGRESS_TIME)
-
-  // Inject any CSS that was generated by mathjax-node
```

Comment:
Does this mean that the HTML+CSS option no longer works?

Reply:
If you look under `Injecting MathJax-created CSS...`, you can find the code to inject the css.